plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c

```diff
@@ -0,0 +1,510 @@
+	info.range = voi->range;
+
+	vaapi_video_info(enc, &info);
+
+	enc->context->profile = profile;
+	enc->context->max_b_frames = bf;
+	enc->context->level = level;
+	enc->context->bit_rate = bitrate * 1000;
+
+	enc->context->width = obs_encoder_get_width(enc->encoder);
+	enc->context->height = obs_encoder_get_height(enc->encoder);
+
+	enc->context->time_base = (AVRational){voi->fps_den, voi->fps_num};
+	enc->context->pix_fmt = obs_to_ffmpeg_video_format(info.format);
+	enc->context->colorspace = info.colorspace == VIDEO_CS_709 ?
+		AVCOL_SPC_BT709 : AVCOL_SPC_BT470BG;
+	enc->context->color_range = info.range == VIDEO_RANGE_FULL ?
+		AVCOL_RANGE_JPEG : AVCOL_RANGE_MPEG;
+
+	if (keyint_sec > 0)
+		enc->context->gop_size = keyint_sec * voi->fps_num / voi->fps_den;
+	else
+		enc->context->gop_size = 120;
+
+	enc->height = enc->context->height;
+
+	info("settings:\n"
+		 "\tdevice:       %s\n"
+		 "\tqp:           %d\n"
+		 "\tquality:      %d\n"
+		 "\tprofile:      %d\n"
+		 "\tlevel:        %d\n"
+		 "\tbitrate:      %d\n"
+		 "\tkeyint:       %d\n"
+		 "\twidth:        %d\n"
+		 "\theight:       %d\n"
+		 "\tb-frames:     %d\n",
+		 device, qp, quality, profile, level,
+		 bitrate, enc->context->gop_size,
+		 enc->context->width, enc->context->height,
+		 enc->context->max_b_frames);
+
+	return vaapi_init_codec(enc, device);
+}
+
+static void vaapi_destroy(void *data)
+{
+	struct vaapi_encoder *enc = data;
+
+	if (!enc->initialized)
+		goto cleanup;
+
+	AVPacket pkt = {0};
+	int r_pkt = 1;
+
+	while (r_pkt) {
+		if (avcodec_encode_video2(enc->context, &pkt, NULL,
+					&r_pkt) < 0)
+			break;
+
+		if (r_pkt)
+			av_free_packet(&pkt);
+	}
+
+cleanup:
+	avcodec_close(enc->context);
+	av_frame_free(&enc->vframe);
+	avpicture_free(&enc->dst_picture);
+	av_buffer_unref(&enc->vaframes_ref);
+	av_buffer_unref(&enc->vadevice_ref);
+	da_free(enc->buffer);
+	bfree(enc->header);
+	bfree(enc->sei);
+
+	bfree(enc);
+}
+
+static void *vaapi_create(obs_data_t *settings, obs_encoder_t *encoder)
+{
+	struct vaapi_encoder *enc;
+	avcodec_register_all();
+
+	enc = bzalloc(sizeof(*enc));
+	enc->encoder = encoder;
+
+	int vaapi_codec = (int)obs_data_get_int(settings, "vaapi_codec");
+
+	if (vaapi_codec == AV_CODEC_ID_H264)
+		enc->vaapi = avcodec_find_encoder_by_name("h264_vaapi");
+
+	enc->first_packet = true;
+
+	blog(LOG_INFO, "---------------------------------");
+
+	if (!enc->vaapi) {
+		warn("Couldn't find encoder");
+		goto fail;
+	}
+
+	enc->context = avcodec_alloc_context3(enc->vaapi);
+	if (!enc->context) {
+		warn("Failed to create codec context");
+		goto fail;
+	}
+
+	if (!vaapi_update(enc, settings))
+		goto fail;
+
+	return enc;
+
+fail:
+	vaapi_destroy(enc);
+	return NULL;
+}
+
+static inline void copy_data(AVPicture *pic, const struct encoder_frame *frame,
+		int height, enum AVPixelFormat format)
+{
+	int h_chroma_shift;
+	int v_chroma_shift;
+	av_pix_fmt_get_chroma_sub_sample(format, &h_chroma_shift, &v_chroma_shift);
```

Comment:
Style: This will need a line-break.

Reply:
It exceeds the line length limit of 80 columns.  A tab is 8-columns, and there are 75 columns of text, which adds up to 83 columns total.  This would be correct:
```
	av_pix_fmt_get_chroma_sub_sample(format, &h_chroma_shift,
			&v_chroma_shift);
```